### PR TITLE
Introduce golangci-lint and fix various issues reported by it.

### PR DESCRIPTION
**Pull request by fnordahl**
Please review commit by commit.

Reported-at: https://launchpad.net/bugs/2072471

### CODE REVIEW

**Review by mkalcok — CHANGES_REQUESTED** [review]
Thank you for introducing this! Static checks and linters make me happy :)

I have just one suggestion inline, otherwise LGTM. (you seem to have used wrong email address to sign the commits.)

File: microovn/cmd/microovnd/main.go
Comment (fnordahl) [reply microovn/cmd/microovnd/main.go:82]: That is a good point, I was unsure given it was not caught by the `unused` linter, but tests do appear to pass without this function, so let's get rid of it.

**Comment by fnordahl** [comment]
> Thank you for introducing this! Static checks and linters make me happy :)
> 
> I have just one suggestion inline, otherwise LGTM. (you seem to have used wrong email address to sign the commits.)

Thanks! Yeah the CLA-check does not handle the Ubuntu member alias, I've raised an issue on that in canonical/has-signed-canonical-cla#47

I'm also working on enabling the `revive` linter, which among other things will give us checks for missing Godoc for public functions.

File: microovn/node/node.go
Comment (fnordahl) [inline microovn/node/node.go:1]: When adding package level Godoc for the `node` package, I have to admit the name appears a bit vague, having our discussion this morning in mind cc @mkalcok @crypticC0der

File: microovn/.golangci.yml
Comment (mkalcok) [inline microovn/.golangci.yml:5]: Could we have a short comment here explaining what's the purpose of this `exclude-rules` statement?

File: microovn/.golangci.yml
Comment (mkalcok) [inline microovn/.golangci.yml:1]: Which linter is doing the GoDoc checking? I skimmed briefly through https://golangci-lint.run/usage/linters/ but I couldn't find a mention of it.

**Review by mkalcok — APPROVED** [review]
Overall LGTM. I left some inline comments to consider.

File: microovn/api/certificates/regenerate_ca.go
Comment (crypticC0der) [inline microovn/api/certificates/regenerate_ca.go:46]: why use %v over %w

File: microovn/api/ovsdb/schema_version.go
Comment (crypticC0der) [inline microovn/api/ovsdb/schema_version.go:169]: i would not use client here as its a package name

File: microovn/api/ovsdb/schema_version.go
Comment (crypticC0der) [inline microovn/api/ovsdb/schema_version.go:178]: i would not use node as its a package name

File: microovn/.golangci.yml
Comment (fnordahl) [reply microovn/.golangci.yml:5]: Yes, it definitively deserves one, on it.

The gist of it is that the default exclude rules disables the GoDoc requirement check, so we need to disable the default exclude rules, which then also disables the one we add back.

File: microovn/ovn/snap.go
Comment (crypticC0der) [inline microovn/ovn/snap.go:61]: reasonable removal of a function :+1:

**Review by crypticC0der — DISMISSED** [review]

File: microovn/api/ovsdb/schema_version.go
Comment (crypticC0der) [reply microovn/api/ovsdb/schema_version.go:169]: :+1:

File: microovn/.golangci.yml
Comment (fnordahl) [reply microovn/.golangci.yml:1]: The `revive` one.

File: microovn/ovn/cmd/cmd.go
Comment (fnordahl) [reply microovn/ovn/cmd/cmd.go:1]: Presumably the rendered docs would retain the fact that it is inside the `ovn` package, but spelling it out does not hurt.

File: microovn/api/ovsdb/schema_version.go
Comment (fnordahl) [reply microovn/api/ovsdb/schema_version.go:169]: Ah, indeed. The linter appears to have no issue with a leading `_` as opposed to a trailing `_`, and I did not think of  any better names for the iterator variables, so I'll go with that

File: microovn/api/certificates/regenerate_ca.go
Comment (fnordahl) [reply microovn/api/certificates/regenerate_ca.go:46]: Simply because the `logger` functions do not support `%w`, which was called out by the linter.

**Comment by fnordahl** [comment]
One more thing I need to fix, my editor defaults to use white space instead of tabs, so apparently I have introduced some discrepancies because of that. Will push an update shortly.

**Comment by crypticC0der** [comment]
#155 merged, conflicts now exist

File: microovn/api/ovsdb/schema_version.go
Comment (crypticC0der) [reply microovn/api/ovsdb/schema_version.go]: there are probably other names that would work here, such as clusterClient, or cClient, or vClient (v to indicate its a variable)

**Comment by mkalcok** [comment]
> One more thing I need to fix, my editor defaults to use white space instead of tabs, so apparently I have introduced some discrepancies because of that. Will push an update shortly.

I'll be honest, I never knew we are using tabs in this project :D :D 

One more thing to consider before we get to merging this. I like the idea of running lint inside snap container to avoid polluting dev machine with unnecessary build dependencies. However it also means that the snap wont build if code does not pass linting, right? This could get annoying during the development, e.g. If I'm just experimenting with something new, I don't want to have to write docs for it right away.

(not a blocker, I just wanted to bring it up.)

**Comment by fnordahl** [comment]
> > One more thing I need to fix, my editor defaults to use white space instead of tabs, so apparently I have introduced some discrepancies because of that. Will push an update shortly.
> 
> I'll be honest, I never knew we are using tabs in this project :D :D

Yeah, `gofmt` appears to mandate it unfortunately, otherwise I would suggest to get rid of them!

> One more thing to consider before we get to merging this. I like the idea of running lint inside snap container to avoid polluting dev machine with unnecessary build dependencies. However it also means that the snap wont build if code does not pass linting, right? This could get annoying during the development, e.g. If I'm just experimenting with something new, I don't want to have to write docs for it right away.
> 
> (not a blocker, I just wanted to bring it up.)

Yeah, this is a concern. I did not find a good answer to it though, there is no "optional" part for snapcraft AFAICT.

Perhaps we could make use of [workshop](https://github.com/canonical/workshop) in the future?

**Comment by mkalcok** [comment]
> To be pragmatic while developing locally, I guess it is trivial to comment out the few lines that runs the linters as part of the build?

Yes, I suspect that this is going to be common workaround. We can get around to changing this if we find it very annoying in the future.

**Comment by mkalcok** [comment]
After the merge conflict resolution, we don't have godocs for some of the exported functions ([example](https://github.com/fnordahl/microovn/blob/787e2c6371f0101e8d329413923e298374f8be14/microovn/node/node.go#L32)) but the CI doesn't seem to be catching it.

**Comment by fnordahl** [comment]
> After the merge conflict resolution, we don't have godocs for some of the exported functions ([example](https://github.com/fnordahl/microovn/blob/787e2c6371f0101e8d329413923e298374f8be14/microovn/node/node.go#L32)) but the CI doesn't seem to be catching it.

Indeed, I accidentally dropped the execution of the `golangci-lint` binary during the rebase :rofl:

**Review by mkalcok — APPROVED** [review]
LGTM, thanks :)

**Pull request merged**
